tests/storage_and_cookie_security_tests.rs

```diff
@@ -0,0 +1,394 @@
+
+        let result = storage.verify_credentials(creds).await.unwrap();
+        assert_eq!(result, VerificationResult::Unauthorized);
+    }
+}
+
+/// Tests for role and permission security
+mod access_control_security_tests {
+    use super::*;
+
+    // Helper function to check if one role supervises another
+    fn supervises<T: AccessHierarchy + PartialEq>(supervisor: &T, subordinate: &T) -> bool {
+        let mut current = supervisor.subordinate();
+        while let Some(role) = current {
+            if role == *subordinate {
+                return true;
+            }
+            current = role.subordinate();
+        }
+        false
+    }
+
+    #[test]
+    fn test_role_hierarchy_security() {
+        // Test that role hierarchy is correctly implemented
+        assert!(supervises(&Role::Admin, &Role::Reporter));
+        assert!(supervises(&Role::Admin, &Role::User));
+        assert!(supervises(&Role::Reporter, &Role::User));
+        
+        // Test that roles don't supervise themselves
+        assert!(!supervises(&Role::User, &Role::User));
+        assert!(!supervises(&Role::Admin, &Role::Admin));
+        
+        // Test inverse relationships
+        assert!(!supervises(&Role::User, &Role::Admin));
+        assert!(!supervises(&Role::User, &Role::Reporter));
+        assert!(!supervises(&Role::Reporter, &Role::Admin));
+    }
+
+    #[tokio::test]
+    async fn test_account_with_multiple_roles() {
+        let account = Account::new(
+            "multiRole@example.com",
+            &[Role::User, Role::Reporter],
+            &[Group::new("users"), Group::new("reporters")],
+        );
+
+        assert_eq!(account.roles.len(), 2);
+        assert!(account.roles.contains(&Role::User));
+        assert!(account.roles.contains(&Role::Reporter));
+        
+        assert_eq!(account.groups.len(), 2);
+        assert!(account.groups.contains(&Group::new("users")));
+        assert!(account.groups.contains(&Group::new("reporters")));
+    }
+
+    #[test]
+    fn test_group_security() {
+        let group1 = Group::new("admin");
+        let group2 = Group::new("admin");
+        let group3 = Group::new("user");
+
+        // Groups with same name should be equal
+        assert_eq!(group1, group2);
+        assert_ne!(group1, group3);
+
+        // Test group name validation
+        let empty_group = Group::new("");
+        assert_eq!(empty_group.name(), "");
+
+        let special_chars_group = Group::new("admin-group_123");
+        assert_eq!(special_chars_group.name(), "admin-group_123");
+    }
+
+    #[tokio::test]
+    async fn test_permission_boundary_enforcement() {
+        use num_enum::{IntoPrimitive, TryFromPrimitive};
+
+        #[derive(Debug, PartialEq, IntoPrimitive, TryFromPrimitive)]
+        #[repr(u32)]
+        enum TestPermission {
+            Read = 0,
+            Write = 1,
+            Delete = 2,
+        }
+
+        let _account = Account::new(
```

Comment:
Whats this account definition for? I do not see the context in this test for it.